### DISCUSSION TITLE
Local multimap file changes race condition?

### DISCUSSION THREAD

**Discussion by ManChicken1911** [rspamd/rspamd#5235] [Q&A]
I notice from time to time when I edit a local file multimap that rspamd jumps on reading it **immediately** upon saving, but I think there might be a race condition or something happening. I use nano as my editor and am editing the file on the same server, on the local disk.

After saving, in the web UI's error log I will often notice a few errors like `empty or invalid key found on line 0` or a regex error on a random line like `cannot parse regexp /(@|\.)aplusc: pattern is not enclosed with /: /(@|\.)aplusc` -- which, that particular rule has a bunch more characters after it in reality (and IS terminated properly), so I feel like it's trying to read the map before it's finished being written to. That rule for example is only about 20% in to the whole file.

Should there perhaps be a small delay imposed when rspamd sees the file change? Even if I just `touch` the file it immediately spits out 5 `rspamd_map_on_stat` status lines before my enter key even fully releases 😁

**Comment by fatalbanana** [comment]
A well-behaved editor should probably not do that, consider using a different one.

Maps should at no point be in a partially-written state. A safe way to update them is to replace the files.
